components/scream/src/physics/share/tests/physics_universal_unit_tests.cpp

```diff
@@ -248,22 +157,60 @@ struct UnitWrap::UnitTest<D>::TestUniversal
       static constexpr Scalar rd     = C::RD;
       static constexpr Scalar inv_cp = C::INV_CP;
       static constexpr Scalar tmelt  = C::Tmelt;
+      static constexpr Scalar ep_2   = C::ep_2;
+      // Gather the machine epsilon for the error tolerance
+      static constexpr Scalar eps = C::macheps;
+      Real tol = 2000*eps;
 
       // Create dummy level data for testing:
       Real pres_top = 200.;
       Real dp       = (p0-pres_top)/(num_levs-1);
 
       // Run tests
+      //========================================================
+      // Test conversion of temperature to potential temperature, and vice versa
+      //========================================================
+      // This function tests both the get_potential_temperature and get_potential_temperature_inv universal conversion functions.
+      //
+      // Inputs:
+      //   T_mid:    An example temperature, K.
+      //   pressure: A pressure value to use for conversion, Pa.
+      //========================================================
+      physics::get_potential_temperature(num_levs,temp,pressure,T_potential);
+      physics::get_potential_temperature_inv(num_levs,T_potential,pressure,T_from_potential);
+      //========================================================
+      // Test calculation of virtual temperature using get_virtual_temperature
+      //========================================================
+      // This function tests the function "get_virtual_temperature".
+      //
+      // Inputs:
+      //   temp is the atmospheric temperature. Units in K.
+      //   qv   is the atmospheric water vapor mass mixing ratio.  Units in kg/kg
+      //========================================================
+      physics::get_virtual_temperature(num_levs,temp,qv,T_virtual);
+      // Check the values from the tests above against expected values over the entire column
       for (int k=0;k<num_levs;++k)
       {
+        // temperature conversion check
+        Real exp_T_pot = physics::get_potential_temperature(temp(k),pressure(k),Smask(true))[0];
```

Comment:
I'm not sure if these tests are meaningful or not. It seems to me that all they are doing is serializing the num_levs loop, and comparing the output of the column version at level k with the single-level output. Even if the test passes, we are not testing much (namely, only that we wrote the TeamThreadRange correctly).

Imho, we should compute a whole column worth of solution, then do all checks on host, using directly REQUIRE (no reduction of nerr). The 'inv' functions can easily be tested by checking that their combo gives back the initial input. For each individual fcn we'd need to compare the output with the math formulation, or test some properties.

To avoid testing the result of a fcn against its mathematical formula, we could do stuff like testing that get_exner(p=p0) returns 1 (up to a tolerance). Or that get_exner(2*p)/get_exner(p) = 2^(rd/cp). But these might be as complicated as testing that get_exner(p)==(p/p0)^(rd/cp)...

In general, btw, we need to be aware that tolerance tests are bound to fail, no matter what. For every tol you pick, there will be some combination of the inputs that can cause ulp errors to cascade and cause an error>tol.

Note: in Homme we implemented a bfb pow, using taylor expansion. Notice that this is _not_ bfb in the sense that f90 matches cxx, but rather that the impl only requires +,-,*,/, which with math optimization disabled lead to bfb results. It worked well for Homme, and maybe we can apply it in scream too, if tol tests start to be a burden.